### DISCUSSION TITLE
Property inside `<For>` does not update

### DISCUSSION THREAD

**Discussion by ramadhanafif** [solidjs/solid#1975] [Q&A]
I'm still new to webdev and to solid.

I currently trying to use `<For>` control flow to render a json data signal.
This question seems similar to #813 but I don't have enough understanding about it, apologies if this is a duplicate.

This is the signal, notice I use `equals:false` because I want to enforce rerender.
```js
const [positions, setPositions] = createSignal(
  [{ device_id: 0xabcd, x: Math.random(), y: 0, z: 0 }],
  { equals: false }, // always rerender on setter call
);
```

and this is the part where it gets called

```jsx
function DummyComponent(props) {
  return (
    <For
      each={positions()}
      fallback={
        <div class="border-black border px-2">
          <h3>no devices</h3>
        </div>
      }
    >
      {(pos, i) => (
        <ul key={pos.device_id} class="border-black border px-2">
          <li>
            device #{i()}: 0x{pos.device_id.toString(16).padStart(4, "0")}
          </li>
          <li>x: {pos.x} </li>
          <li>y: {pos.y} </li>
          <li>z: {pos.z} </li>
        </ul>
      )}
    </For>
  );
}
```

The part where the signal gets updated has been confirmed to work correctly using `createEffect()` + `console.log()`

## The Issue
As you can see, on the `createSignal()` part above, the object inside signal has an x, y, and z values. The issue is whenever I change the x y or z value, the signal does not gets updated.

Take a look at the function below, where it will randomize x y z value, and then call the signal setter. 

```jsx
function randomPos() {
  setPositions((current) => {
    current.forEach((element, index) => {
      console.log("randomize position");
      element.x = Math.random();
      element.y = Math.random();
      element.z = Math.random();
    });
    return current;
  });
}
```

I expect the `randomPos()` function will rerender the section inside `<For>`, but in reality it does not!

However, if I add items to the signal list, it will update.

Can someone kindly give me a simple technical explanation why does it behave this way, and what are the possible solution to this issue?

**Comment by ramadhanafif** [comment]
Update: so I tried to do a deep copy for the signal, and it works. 
Now the `randomPos()` function looks like this:

```jsx
function changePos() {
  setPositions((current) => {
    current.forEach((element, index) => {
      console.log("randomize position");
      element.x = Math.random();
      element.y = Math.random();
      element.z = Math.random();
    });
    return JSON.parse(JSON.stringify(current)); // Deep-copy trick
  });
}
```

Feels like stupid and unoptimized way to solve this issue. Does anyone has a better way?

**Reply by trusktr** [reply]
Signals are a single value, they don't track anything inside of them, only the top-level value.

If you replace `createSignal` with `createMutable` (import `from 'solid-js/store'`), then what you tried will work. A mutable is an object where you can modify any level deep, and effects can read any level deep.

**Reply by trusktr** [reply]
Also avoid using JSON like that, that'll lower your performance by a lot.

```js
function changePos() {
  setPositions((current) => {
    current.forEach((element, index) => {
      console.log("randomize position");
      element.x = Math.random();
      element.y = Math.random();
      element.z = Math.random();
    });
    return [...current] // much faster
  });
}
```

**Reply by trusktr** [reply]
Also look at `createStore`, it is like a mutable where an effect can read any level deep, but it requires using a separate setter function to set values any level deep. This is useful if you want to pass the store around and want it to be read-only, while you keep the setter private (f.e. do not export the setter from your module, or similar).
